### ISSUE TITLE
Rails 6 parallel testing ActiveSupport not calling with_info_handler method, causing minitest-hooks plugin to raise NoMethodError

### ISSUE THREAD

**Issue by kazooiebombchu** [rails/rails#35495]
### Steps to reproduce
* Create new rails project with new rails that has parallel testing enabled by default
* Install `minitest-hooks` minitest plugin gem
* Add `include Minitest::Hooks` to file `test/test_helper.rb` in `ActiceSupport::TestCase` class
* Add one dummy test and also add `before_all` method to test class.
* Run `rails test`

I also created [github repo](https://github.com/kazooiebombchu/testing-active-support-bug) to replicate this issue. Clone the repo and run `rails test`.

### Expected behavior
Tests should be run without problems.

### Actual behavior
Running tests raises NoMethodError. I made issue about this to [minitest-hooks](https://github.com/jeremyevans/minitest-hooks/issues/16) but author mentioned this being a bug in ActiveSupport not calling `with_info_handler` method around of calling `run_one_method` which minitest-hooks needs. Is this bug or intended behavior?

Here is the stacktrace from the error:
```
Traceback (most recent call last):
	19: from -e:1:in `<main>'
	18: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	17: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	16: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application/boot.rb:19:in `<top (required)>'
	15: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:135:in `run'
	14: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:135:in `loop'
	13: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:141:in `block in run'
	12: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:171:in `serve'
	11: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:171:in `fork'
	10: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/minitest-5.11.3/lib/minitest.rb:63:in `block in autorun'
	 9: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/minitest-5.11.3/lib/minitest.rb:133:in `run'
	 8: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:70:in `start'
	 7: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:70:in `map'
	 6: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:70:in `each'
	 5: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:70:in `times'
	 4: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:71:in `block in start'
	 3: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:71:in `fork'
	 2: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.0.beta2/lib/active_support/testing/parallelization.rb:82:in `block (2 levels) in start'
	 1: from /home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/minitest-5.11.3/lib/minitest.rb:960:in `run_one_method'
/home/kazooie/.rbenv/versions/2.6.0-rc2/lib/ruby/gems/2.6.0/gems/minitest-hooks-1.5.0/lib/minitest/hooks/test.rb:56:in `new': undefined method `name=' for nil:NilClass (NoMethodError)
```

### System configuration
**Rails version**: 6.0.0.beta2

**Ruby version**: 2.6.0rc2 (2018-12-15 trunk 66408) [x86_64-linux]

**Comment by eileencodes** [comment]
@kazooiebombchu I think it's just an oversight on our side. Do you want to send a PR to fix it?

**Comment by kazooiebombchu** [comment]
yeah I can try to fix this. I'm not familiar rails codebase but I think I manage.

**Comment by kazooiebombchu** [comment]
I actually need to ask to get knowledge if this is even worth of fixing. At the moment I managed to fix the problem but with this [diff](https://github.com/kazooiebombchu/rails/commit/e21c243dd6da6f85d4abfb4415dfbd151b172d38).

However, we are using [minitest-hooks](https://github.com/jeremyevans/minitest-hooks) minitest plugin which allows us to run `before_all` and `after_all` methods before test  class. This is used to setup heavy database setup before running tests on certain test classes. The database setup is done in `before_all` and reverted back original in `after_all`.

Right now parallel testing only executes one test at the time, which makes this setup quite useless. Forked process always instantiates class with `new` and runs one test method. This is ideal if only `setup` method is used.

Now with the diff I provided following will happen based on how minitest-hooks is implemented: before tests are pushed to the DRb queue, every pushed test class get it's `before_all` and `after_all` executed for nothing. This happens because minitest-hooks [overrides](https://github.com/jeremyevans/minitest-hooks/blob/master/lib/minitest/hooks/test.rb#L64) `with_info_handler` method that will run above methods. When forked process pops job from the DRb queue, `before_all` and `after_all` will get executed yet again for every test method there is on the class. This makes it basically same as normal `setup` method.

So I was wondering is this even worth of fixing? Do we need setup methods like this in general or is it bad design? At least what I've coded Python UnitTests also have this feature. I can think of couple ideas of top of my head how to fix this if needed:
* Each forked process buffers methods for one test class and executes them in one patch later.
* Each forked process is passed only with test class which then takes care of running all it's test methods with setup and tear down.

Both above would need overriding method `Runnable#run` on `ActiveSupport#TestCase` which will push test methods to the queue without calling `with_info_handler` which is not needed.

I guess the main point to ask is does rails people want setup methods at all in testing or are they considered bad practice?

**Comment by kazooiebombchu** [comment]
I quickly did what I meant above. [This](https://github.com/kazooiebombchu/rails/commit/ec742c9badbe350d6e5e3d9fbb56d8abaca2a9e1) ugly hack is just a proof of concept.

It buffers all test methods for one test class and pushes all of them at once. This causes one process to execute all test methods for one test class with correct behavior with `before_all` and `after_all` hooks.

Comments are welcome :)

**Comment by eileencodes** [comment]
This is a tough one. We didn't design parallel testing to do each test individually by accident. We at GitHub have a per suite parallelizer and it's not as efficient as a per-test parallelizer. The only way I think we'd accept doing it per-suite is if it was opt-in rather than the default and only behavior.

I'm not going to tell you it's _bad design_ to use the before_all/after_all because we do it as well. However it definitely can cause problems and I'm not sure how supportable this is given your investigation.

Have you used other concurrent testing gems? Do you know how those work with the minitest-hooks gem?

**Comment by kazooiebombchu** [comment]
Yeah if test suites are not designed evenly then overall testing time will take longer. We currently use [parallel_tests](https://github.com/grosser/parallel_tests) gem. So far what I understand and have tested it divides tests to processes by suites and executes them. So befofe_all and after_all gets executed only once for each suite and it works correctly.

Why we're changing away from that is that we're hoping to use rails own test runners instead. Now at the moment with rails 6.0.0.beta2 running tests will not work at all if using minitest-hooks.

To me both kinda feels correct at this point. However I think calling tests `with_info_handler` method would be okay fix to implement at this point so tests are even runnable, even `before_all` gets executed for each test method. Then just put some information about this somewhere.

**Comment by eileencodes** [comment]
> However I think calling tests with_info_handler method would be okay fix to implement at this point so tests are even runnable, even before_all gets executed for each test method.

Yes I think this is fine. I missed your first commit that you linked to and only saw the second fix you wrote later on. You can open a PR with the first one (https://github.com/kazooiebombchu/rails/commit/e21c243dd6da6f85d4abfb4415dfbd151b172d38). Thanks!

**Comment by kazooiebombchu** [comment]
Any news on merging above PR? Or is there something I'm missing?

**Comment by eileencodes** [comment]
I need to test out a few things first. It's on the Rails 6.0 milestone so don't worry it won't get released without this change unless something major blocks it.

**Comment by kazooiebombchu** [comment]
Okay thanks! Let me know if you need any changes or something related to this. I'm more than happy to help.

**Comment by y-yagi** [comment]
Fixed with 758ba117a008b6ea2d3b92c53b6a7a8d7ccbca69
